test/markup/cpp/templates.txt

```diff
@@ -0,0 +1,104 @@
+template <template <typename X> class T> void f();
+
+template class A::B<1, 2> {};
+
+template void A::f(int);
+
+template <class T, std::size_t N = sizeof(T), class K = T>
+ostream& operator<< (ostream& os, const skg::Triplet<T> pt) ;
+void other_stuff_that_isnt_colored();
+
+
+template<class T>
+class My_vector { /* ... */ };
+
+template<class T = void>
+struct My_op_functor { /* ... */ };
+
+template<typename... Ts>
+class My_tuple { /* ... */ };
+
+
+template<class> class My_vector;
+template<class = void> struct My_op_functor;
+template<typename...> class My_tuple;
+
+
+template<typename T> concept C1 = true;
+template<typename... Ts> concept C2 = true; // variadic concept
+template<typename T, typename U> concept C3 = true;
+ 
+template<C1 T> struct s1;         // constraint-expression is C1<T>
+template<C1... T> struct s2;      // constraint-expression is (C1<T> && ...)
+template<C2... T> struct s3;      // constraint-expression is (C2<T> && ...)
+template<C3<int> T> struct s4;    // constraint-expression is C3<T, int>
+template<C3<int>... T> struct s5; // constraint-expression is (C3<T, int> && ...)
+
+
+template<typename K, typename V, template<typename> typename C = my_array>
+class Map {};
+
+
+template<class B>
+template<class C>
+void A<B>::g(C) {}
+
+
+void g() {
+    f<int()>();
+}
+
+
+template<int (&pa)[5]> struct W {};
+template<void (*pf)(int)> struct A {};
+
+
+template<class T, const char* p> class X {};
+
+
+template<typename T> class A { int x; }; // primary template
+template<typename T> class A<T*> { long x; }; // partial specialization
+
+
+template<typename T> struct eval; // primary template 
+ 
+template<template<typename, typename...> class TT, typename T1, typename... Rest>
+struct eval<TT<T1, Rest...>> {}; // partial specialization of eval
+
+
+eval<A<int>> eA;
+eval<B<int, float>> eB;
+eval<C<17>> eC;
+eval<D<int, 17>> eD;
+eval<E<int, float>> eE;
+
+
+template <template <auto> class> void FA();  // note: C++17
```

Comment:
This isn't a template match because it has a ` ` after template so the `<` will never match because the rule will end when it hits a space (which doesn't match any of your TEMPLATE rules...

Adding a space eater to your begin matcher for `<` should help.

Reply:
Are you using the developer tool?  This kind of thing is a lot easier to see with the visual aids of block nesting.

Reply:
Why would whitespace be an issue here? It works fine for `TEMPLATE_DECLARATION`. That said, nesting templates are actually an issue I had forgotten to comment on: how would I do that? I’ve got a circular dependency and I can’t use `self`, since the mode is nested deeper (i.e. `TEMPLATE_DECLARATION` does *not* contain `TEMPLATE_DECLARATION` — but it contains a mode which should contain `TEMPLATE_DECLARATION`).

In fact, I do use the developer tool, and here’s what I’m currently seeing for this input (I’ve added additional class names for debugging purposes):

<img width="1008" alt="Screenshot 2020-10-04 at 14 47 18" src="https://user-images.githubusercontent.com/38931/95017316-8f85e800-0650-11eb-9f6c-a34b6e2c010f.png">

Ideally we’d want this to have another `tpl-decl` inside the `tpl-params`.

Reply:
Ah I see now, you don't have a class (which you shouldn't) - so I wasn't seeing it because I was only looking at the HTML. :-)  You have an `end` rule so it works... I missed.  If you didn't have and `end` the mode would immediately terminate if it could find no children.

> That said, nesting templates are actually an issue I had forgotten to comment on: how would I do that? I’ve got a circular dependency and I can’t use self, since the mode is nested deeper 

You'd have to do it manually (which will limit depth)... Or rejigger the rules so that say your template rule was say `template<` to `>`...  so that you could then use self...

Reply:
The two nested modes are purely so that I could start with a (highlighted) keyword (`template`), followed by a non-keyword character. I initially had `begin: /template\s*</` but (I thought) that didn’t highlight `template` as a keyword.

But I can’t reproduce this now: it *does* highlight `template` correctly, so I guess I can simplify that nested mode and make it recursive.

Reply:
> that didn’t highlight template as a keyword. ... But I can’t reproduce this now:

It should indeed IF it's in the `keyword` object at the same level... `beginKeywords` hooks this up automatically... if you're writing the rule by hand you'd need both `begin` and `keyword`.